runtime-sdk/src/storage/confidential.rs

```diff
@@ -0,0 +1,617 @@
+fn unpack_nonce_slice<'a>(packed: &'a [u8]) -> Option<(&'a Nonce, &'a [u8])> {
+    if packed.len() <= NONCE_SIZE {
+        return None;
+    }
+    let nonce_ref: &'a Nonce = packed[..NONCE_SIZE]
+        .try_into()
+        .expect("nonce size mismatch");
+    Some((nonce_ref, &packed[NONCE_SIZE..]))
+}
+
+/// Errors emitted by the confidential store.
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("corrupt key")]
+    CorruptKey,
+
+    #[error("corrupt value")]
+    CorruptValue,
+
+    #[error("decryption failure: {0}")]
+    DecryptionFailure(anyhow::Error),
+}
+
+/// A key-value store that encrypts all content with DeoxysII.
+pub struct ConfidentialStore<S: Store> {
+    inner: S,
+    deoxys: deoxysii::DeoxysII,
+    base_value_prefix: Vec<u8>,
+    nonce_counter: usize,
+    nonce_key: Zeroizing<Vec<u8>>,
+}
+
+impl<S: Store> ConfidentialStore<S> {
+    /// Create a new confidential store with the given keypair.
+    pub fn new_with_key(inner: S, key: [u8; KEY_SIZE], value_context: &[&[u8]]) -> Self {
+        let actual_key = Zeroizing::new(key);
+
+        // Derive a nonce key for nonces used to encrypt storage keys in the store:
+        // nonce_key = KDF(key)
+        let mut kdf = Kdf::new_from_slice(b"oasis-runtime-sdk/confidential-store: nonce key")
+            .expect("Hmac::new_from_slice");
+        kdf.update(&key);
+        let mut derived = kdf.finalize().into_bytes();
+        // Try to destroy as much of the bytes as possible; there's
+        // no neat way to get from kdf output to a Vec<u8> without copying a lot.
+        let derived = Zeroizing::new(derived.iter_mut());
+
+        ConfidentialStore {
+            inner,
+            deoxys: deoxysii::DeoxysII::new(&actual_key),
+            base_value_prefix: value_context.concat(),
+            nonce_counter: 0,
+            nonce_key: Zeroizing::new(derived.as_slice().to_vec()),
+        }
+    }
+
+    fn pack_nonce_slice(&self, nonce: &Nonce, slice: &[u8]) -> Vec<u8> {
+        let mut ret = Vec::with_capacity(nonce.len() + slice.len());
+        ret.extend_from_slice(nonce);
+        ret.extend_from_slice(slice);
+        ret
+    }
+
+    fn make_key(&self, plain_key: &[u8]) -> (Nonce, Vec<u8>) {
+        // The nonce used to encrypt storage keys is derived from a combination
+        // of a base nonce key (derived from the encryption key) and the
+        // incoming plaintext storage key:
+        // nonce = Trunc(NONCE_SIZE, H(nonce_key || plain_key))
+        let mut nonce = [0u8; NONCE_SIZE];
+
+        let mut nonce_src = self.nonce_key.clone();
+        nonce_src.extend_from_slice(plain_key);
+
+        let hash = Hash::digest_bytes(&nonce_src);
+        nonce.copy_from_slice(hash.truncated(NONCE_SIZE));
+
+        let enc_key = self.deoxys.seal(&nonce, plain_key.to_vec(), vec![]);
+        let key = self.pack_nonce_slice(&nonce, &enc_key);
+        (nonce, key)
+    }
+
+    fn make_value(&mut self, plain_value: &[u8]) -> (Nonce, Vec<u8>) {
```

Comment:
This seems kind of overkill to mask "the code accessed the same key", but I can't think of a nice way to handle this off the top of my head.